pyhf/__init__.py

```diff
@@ -229,54 +156,28 @@ def add_modifier(self, name, npars, modifier, suggested_init, suggested_bounds):
             'suggested_init': suggested_init,
             'suggested_bounds': suggested_bounds
         }
-        if is_constraint:
+        if is_constrained:
             self.auxdata += self.modifier(name).auxdata
             self.auxdata_order.append(name)
         return True
 
     def add_modifier_from_def(self, channel, sample, modifier_def):
-        if modifier_def['type'] == 'normfactor':
-            modifier = None  # no object for factors
-            self.add_modifier(name=modifier_def['name'],
-                                modifier=modifier,
-                                npars=1,
-                                suggested_init=[1.0],
-                                suggested_bounds=[[0, 10]])
-        if modifier_def['type'] == 'shapefactor':
-            modifier = None  # no object for factors
-            self.add_modifier(name=modifier_def['name'],
-                                modifier=modifier,
-                                npars=len(sample['data']),
-                                suggested_init   =[1.0] * len(sample['data']),
-                                suggested_bounds=[[0, 10]] * len(sample['data'])
-                        )
-        if modifier_def['type'] == 'shapesys':
-            # we reserve one parameter for each bin
-            modifier = shapesys_constraint(sample['data'], modifier_def['data'])
-            self.add_modifier(
-                name=modifier_def['name'],
-                npars=len(sample['data']),
-                suggested_init=[1.0] * len(sample['data']),
-                suggested_bounds=[[0, 10]] * len(sample['data']),
-                modifier=modifier,
-            )
-        if modifier_def['type'] == 'normsys':
-            modifier = normsys_constraint()
-            self.add_modifier(name=modifier_def['name'],
-                         npars=1,
-                         modifier=modifier,
-                         suggested_init=[0.0],
-                         suggested_bounds=[[-5, 5]])
-            self.modifier(modifier_def['name']).add_sample(channel, sample, modifier_def['data'])
-        if modifier_def['type'] == 'histosys':
-            modifier = histosys_constraint()
-            self.add_modifier(
-                modifier_def['name'],
-                npars=1,
-                modifier=modifier,
-                suggested_init=[1.0],
-                suggested_bounds=[[-5, 5]])
-            self.modifier(modifier_def['name']).add_sample(channel, sample, modifier_def['data'])
+        # get the class associated with the modifier
+        modifier_cls = modifiers.registry.get(modifier_def['type'], None)
+        if modifier_cls is None:
+            raise RuntimeError('shared systematic not implemented yet (processing {0:s})'.format(modifier_def['type']))
+
+        # use inspect to figure out how to call the modifier
+        args = inspect.getargspec(modifier_cls.__init__).args
```

Comment:
ideally, i'd like to try to not use `inspect` if we factor the modifiers out we should try to find a uniform API

Reply:
yeah, this is a monkey-patch until I get around to that part.

Reply:
This is now fixed up.